server/blob/blob-deduplicating/src/main/scala/GC.scala

```diff
@@ -0,0 +1,166 @@
+
+  def next: Generation
+  def next(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.next)
+
+  /**
+   * List all generations the GC is able to collect
+   */
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    Generation.range(next, targetedGeneration.previous(GC.temporization)).toSet
+
+  def <(that : Generation): Boolean = compareTo(that) < 0
+  def <=(that : Generation): Boolean = compareTo(that) <= 0
+  def >(that : Generation): Boolean = compareTo(that) > 0
+  def >=(that : Generation): Boolean = compareTo(that) >= 0
+}
+
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = Generation(id - 1)
+
+  override def next: Generation = copy(id + 1)
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 1
+    case that : ValidGeneration => id.compareTo(that.id)
+  }
+
+}
+
+case object NonExistingGeneration extends Generation {
+  override def previous: Generation = NonExistingGeneration
+
+  override def next: Generation = Generation.first
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 0
+    case _ : ValidGeneration => -1
+  }
+}
+
+object Generation {
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      NonExistingGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+
+  def range(start : Generation, end : Generation) : Seq[Generation] = (start, end) match {
+    case (NonExistingGeneration, NonExistingGeneration) => Seq(NonExistingGeneration)
+    case (ValidGeneration(_), NonExistingGeneration) => Nil
+    case (NonExistingGeneration, ValidGeneration(id)) =>  NonExistingGeneration +: (0L to id).map(Generation.apply)
+    case (ValidGeneration(id1), ValidGeneration(id2)) => (id1 to id2).map(Generation.apply)
+  }
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(), NonExistingGeneration)
+}
+
+case class ExternalID(id: String) // TODO
+
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
+}
+
+case class Reference(externalId: ExternalID, blobId: BlobId, generation: Generation) extends Event {
+  override def blob: BlobId = blobId
+}
+
+case class Deletion(generation: Generation, reference: Reference) extends Event {
+  override def blob: BlobId = reference.blob
+  override def externalId: ExternalID = reference.externalId
+}
+
+object Events {
+  def getLastGeneration(events: Seq[Event]): Generation = events.map(_.generation).maxOption
+    .getOrElse(Generation.first)
+
+}
+
+case class Report(iteration: Iteration, blobsToDelete: Set[(Generation, BlobId)])
```

Comment:
I understand most of the concepts that are expressed by these case classes but I think that having some javadoc attached to each one would be a very great thing for other people to understand the code